sound/soc/sof/intel/hda.c

```diff
@@ -1202,7 +1202,8 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
 		}
 		/* Found if all Slaves are checked */
 		if (i == hdev->info.count || !link->num_adr)
-			break;
+			if (!mach->machine_quirk || mach->machine_quirk(hdev->sdw))
+				break;
```

Comment:
if the machine quirk results a non-NULL value, that means the machine is valid.
If the machine quirk returns NULL, we want to keep looking for an alternate entry in the table.

You have to look at the table, we have the VB first then VA. If we have all the peripherals listed with the expected _ADR, the machine quirk will either stop on the VB entry or skip to use VA.

Reply:
I mean, in context of the level up comment..
We have found a match if I read it right and now we are going to run an 'are you sure' query on it?

Reply:
yes, the initial match is really only a partial match based on _ADR information, but we need a second level filter to disambiguate identical configurations used for different types of hardware.
If this looks like a mess it's because it is. I've share suggested text to be added to the SDCA spec to prevent this from happening, but for now it is what it is.

Reply:
I added a "machine_check" callback to solve this return value problem